src/adhocracy_core/adhocracy_core/auditing/__init__.py

```diff
@@ -0,0 +1,116 @@
+"""Auditlog of events stored in a ZODB database."""
+import transaction
+import json
+import substanced.util
+
+from pyramid.traversal import resource_path
+from BTrees.OOBTree import OOBTree
+from datetime import datetime
+from logging import getLogger
+from collections import namedtuple
+from adhocracy_core.utils import get_user
+from adhocracy_core.sheets.principal import IUserBasic
+from adhocracy_core.utils import get_sheet_field
+
+
+logger = getLogger(__name__)
+
+AuditEntry = namedtuple('AuditEntry', ['name', 'payload'])
+
+RESOURCE_CREATED = 'resourceCreated'
+RESOURCE_MODIFIED = 'resourceModified'
+
+
+class AuditLog(OOBTree):
+
+    """An Auditlog composed of entries."""
+
+    def add(self, name, **kw):
+        """ Add a record the audit log.
+
+        ``_name`` should be the event name,
+        ``_oid`` should be an object oid or ``None``, and ``kw`` should be a
+        json-serializable dictionary.
+        """
+        payload = json.dumps(kw)
+        self[datetime.utcnow()] = AuditEntry(name, payload)
+
+
+def get_auditlog(context):
+    """Return the auditlog. Create one if none exits."""
+    auditlog = substanced.util.get_auditlog(context)
+    if auditlog is None:
+        _set_auditlog(context)
+        transaction.commit()
+        auditlog = substanced.util.get_auditlog(context)
+        # auditlog can still be None after _set_auditlog if not audit
+        # conn has been configured
+        if auditlog is not None:
+            logger.info('Auditlog created')
+        return auditlog
+    return auditlog
+
+
+def _set_auditlog(context):
+    """Set an auditlog for the context."""
+    conn = context._p_jar
+    try:
+        auditconn = conn.get_connection('audit')
+    except KeyError:
+        return
+    root = auditconn.root()
+    if 'auditlog' not in root:
+        auditlog = AuditLog()
+        root['auditlog'] = auditlog
+
+
+def log_auditevent(context, name, **kw):
+    """Add a an audit entry to the audit database.
+
+    The audit database is created if missing. If the zodbconn.uri.audit
+    value is not specified in the config, auditing does not happen.
+    """
+    auditlog = get_auditlog(context)
+    if auditlog is not None:
+        auditlog.add(name, **kw)
+
+
+def audit_changes_callback(request, response):
+    """Add audit entries to the auditlog when the data is changed."""
```

Comment:
I did not understand at once what this function does, so I would add the hind "finished_callback" to the function name or the docstring.